libraries/bot-applicationinsights/src/test/java/com/microsoft/bot/applicationinsights/TelemetryWaterfallTests.java

```diff
@@ -0,0 +1,396 @@
+        // so instead of calling the same method, it calls a method with less parameters.
+        // In this particular test, WaterfallDialog's beginDialog calls the method with only two parameters
+        verify(telemetryClient, times(4)).trackEvent(
+            anyString(),
+            anyMap()
+        );
+    }
+
+    @Test
+    public void waterfallWithActionsNull() {
+        BotTelemetryClient telemetryClient = Mockito.mock(BotTelemetryClient.class);
+        WaterfallDialog waterfall = new WaterfallDialog("test", null);
+        waterfall.setTelemetryClient(telemetryClient);
+        try {
+            waterfall.addStep(null);
+            Assert.fail();
+        } catch (IllegalArgumentException ex) {
+            ex.printStackTrace();
+        }
+    }
+
+    @Test
+    public void ensureEndDialogCalled() {
+        ConversationState convoState = new ConversationState(new MemoryStorage());
+        TestAdapter adapter = new TestAdapter(TestAdapter.createConversationReference("ensureEndDialogCalled", "User1", "Bot"))
+            .use(new AutoSaveStateMiddleware(convoState));
+
+        StatePropertyAccessor<DialogState> dialogState = convoState.createProperty("dialogState");
+        DialogSet dialogs = new DialogSet(dialogState);
+        HashMap<String, Map<String, String>> saved_properties = new HashMap<>();
+        final Integer[] counter = {0};
+
+        // Set up the client to save all logged property names and associated properties (in "saved_properties").
+        BotTelemetryClient telemetryClient = new BotTelemetryClient() {
+            @Override
+            public void trackAvailability(String name, OffsetDateTime timeStamp, Duration duration, String runLocation, boolean success, String message, Map<String, String> properties, Map<String, Double> metrics) {
+
+            }
+
+            @Override
+            public void trackDependency(String dependencyTypeName, String target, String dependencyName, String data, OffsetDateTime startTime, Duration duration, String resultCode, boolean success) {
+
+            }
+
+            @Override
+            public void trackEvent(String eventName, Map<String, String> properties, Map<String, Double> metrics) {
+                StringBuilder sb = new StringBuilder(eventName).append("_").append(counter[0]++);
+                saved_properties.put(sb.toString(), properties);
+            }
+
+            @Override
+            public void trackException(Exception exception, Map<String, String> properties, Map<String, Double> metrics) {
+
+            }
+
+            @Override
+            public void trackTrace(String message, Severity severityLevel, Map<String, String> properties) {
+
+            }
+
+            @Override
+            public void trackDialogView(String dialogName, Map<String, String> properties, Map<String, Double> metrics) {
+
+            }
+
+            @Override
+            public void flush() {
+
+            }
+        };
```

Comment:
Why this? This is [mocked](https://github.com/microsoft/botbuilder-dotnet/blob/main/tests/Microsoft.Bot.Builder.ApplicationInsights.Tests/TelemetryWaterfallTests.cs#L108)

Reply:
Maybe we need to mock only the `trackEvent` instead of this

Reply:
I used as example how they tested the [Storage](https://github.com/microsoft/botbuilder-java/blob/4dd06e088f45616ef5bd1aae140e6726bd682332/libraries/bot-builder/src/test/java/com/microsoft/bot/builder/BotStateTests.java#L44-L63) class.

Reply:
Check about the `when` and `return` of Mockito. But yes, good approach to review another implementation

Reply:
`Mockito.doAnswer` was the correct answer

Reply:
😍